### PR TITLE
MAILET-110 Refactor ReplaceContent

### PR DESCRIPTION
**Pull request by aduprat**
I split this review in two commit as it may be more readable.
I'm not really proud of new class names.

I'm looking at the repeat feature, as i think it's buggy.

### CODE REVIEW

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java [hunk 1]
```diff
@@ -88,263 +94,287 @@
...
+        if (Integer.valueOf(getInitParameter("debug", "0")) == 1) {
+            return true;
+        }
+        return false;
+    }
+
+    private ReplaceConfig initPatterns() throws MailetException {
+        try {
+            ReplaceConfig.Builder builder = ReplaceConfig.builder();
+            initSubjectPattern(builder);
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: I'd rather write `builder().addAllSubjectReplacingUnits(subjectPattern())` , etc

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java [hunk 2]
```diff
@@ -88,263 +94,287 @@
...
-    
+
+    private static void ensurePatternIsValid(String pattern) throws MailetException {
+        if (pattern.length() < 2 || (!pattern.startsWith("/") && !pattern.endsWith("/"))) {
+            throw new MailetException("Invalid expression: " + pattern);
+        }
+    }
+
+    private static ReplacingPattern getPattern(String line) throws MailetException {
+        String[] pieces = StringUtils.split(line, "/");
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: parts

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java [hunk 3]
```diff
@@ -88,263 +94,287 @@
...
+    }
+
+    private void initBodyPatternFile(ReplaceConfig.Builder builder) throws MailetException, IOException {
+        String filePattern = getInitParameter(PARAMETER_NAME_BODY_PATTERNFILE);
+        if (filePattern != null) {
+            builder.addAllBodyReplacingUnits(getPatternsFromFileList(filePattern));
+        }
+    }
+
+    protected static List<ReplacingPattern> getPatternsFromString(String pattern) throws MailetException {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: this method deserves its own class

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java [hunk 4]
```diff
@@ -88,263 +94,287 @@
...
+            throw new MailetException("Invalid expression: " + pattern);
+        }
+    }
+
+    private static ReplacingPattern getPattern(String line) throws MailetException {
+        String[] pieces = StringUtils.split(line, "/");
+        if (pieces.length < 3) {
+            throw new MailetException("Invalid expression: " + line);
+        }
+        return new ReplacingPattern(Pattern.compile(pieces[0], extractOptions(pieces[2])),
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: you could name the parts by using constant like that : `static final int PATTERN = 0;`

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        this.debug = debug;
+        this.logOwner = logOwner;
+    }
+
+    public String applyPatterns(List<ReplacingPattern> patterns, String text) {
+        for (ReplacingPattern replacingPattern : patterns) {
+            boolean changed;
+            do {
+                changed = false;
+                Matcher matcher = replacingPattern.getMatcher().matcher(text);
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: please extract the replace of one occurrence

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                    text = replaced;
+                    changed = true;
+                }
+            } while (replacingPattern.isRepeat() && changed);
+        }
+        
+        return text;
+    }
+
+    public void applyPattern(Mail mail, ReplaceConfig replaceConfig, Optional<Charset> charset) throws MailetException {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: which pattern ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    private boolean applySubjectReplacingUnits(Mail mail, ReplaceConfig replaceConfig, Optional<Charset> charset) throws MessagingException {
+        if (!replaceConfig.getSubjectReplacingUnits().isEmpty()) {
+            String subject = applyPatterns(replaceConfig.getSubjectReplacingUnits(), 
+                    Strings.nullToEmpty(mail.getMessage().getSubject()));
+            if (charset.isPresent()) {
+                mail.getMessage().setSubject(subject, charset.get().name());
+            }
+            else {
+                mail.getMessage().setSubject(subject);
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: I think you should pass the previous charset here

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            return true;
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: you don't know if anything changed

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        
+        return text;
+    }
+
+    public void applyPattern(Mail mail, ReplaceConfig replaceConfig, Optional<Charset> charset) throws MailetException {
+        try {
+            boolean subjectChanged = applySubjectReplacingUnits(mail, replaceConfig, charset);
+            boolean contentChanged = applyBodyReplacingUnits(mail, replaceConfig, charset);
+
+            if (charset.isPresent() && !contentChanged) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: why would you want to change the charset if nothing changed ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+import com.google.common.base.Optional;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+
+public class PatternExtractor {
+
+    public List<ReplacingPattern> getPatternsFromString(String pattern) throws MailetException {
+        String patternProcessing = pattern.trim();
+        ensurePatternIsValid(patternProcessing);
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: assertPatternSurroundedWithSlashes

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        patternProcessing = patternProcessing.substring(1, patternProcessing.length() - 1);
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: trim surroundingSlashes

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        String[] patternArray = StringUtils.split(patternProcessing, "/,/");
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: this and the previous line should be in a method extractPatternParts

Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: use Guava Strings for that

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        
+        ImmutableList.Builder<ReplacingPattern> patternList= ImmutableList.builder();
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: space missing before  `=`

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        for (String aPatternArray : patternArray) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: patternArrays

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            throw new MailetException("Invalid expression: " + line);
+        }
+        return new ReplacingPattern(Pattern.compile(pieces[0], extractOptions(pieces[2])), 
+                extractRepeat(pieces[2]), 
+                unescapeSubstitutions(pieces[1]));
+    }
+
+    private int extractOptions(String optionsAsString) {
+        int options = 0;
+        if (optionsAsString.indexOf('i') >= 0) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: contains ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            options += Pattern.CASE_INSENSITIVE;
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: `|=` ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+            options += Pattern.MULTILINE;
+        }
+        if (optionsAsString.indexOf('s') >= 0) {
+            options += Pattern.DOTALL;
+        }
+        return options;
+    }
+
+    private boolean extractRepeat(String flagsAsString) {
+        if (flagsAsString.indexOf('r') >= 0) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: `return flagsAsString.contains('r');`

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        return false;
+    }
+
+    private String unescapeSubstitutions(String substitutions) {
+        String unescaped = substitutions;
+        if (unescaped.contains("\\r")) {
+            unescaped = unescaped.replaceAll("\\\\r", "\r");
+        }
+        if (unescaped.contains("\\n")) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: why don't you search for `\\\\n` ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        if (unescaped.contains("\\t")) {
+            unescaped = unescaped.replaceAll("\\\\t", "\t");
+        }
+        return unescaped;
+    }
+
+    public List<ReplacingPattern> getPatternsFromFileList(String filepar) throws MailetException, IOException {
+        ImmutableList.Builder<ReplacingPattern> patternList= ImmutableList.builder();
+        String[] files = filepar.split(",");
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: Guava spliter handle the triming I guess

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+        for (String file : files) {
+            file = file.trim();
+            String charset = null;
+            int charsetOffset = file.lastIndexOf('?');
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: needs some abstraction, it's too raw right now

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+            }
+        }
+        return patternList.build();
+    }
+
+    private List<ReplacingPattern> getPatternsFromStream(InputStream stream, String charset) throws MailetException, IOException {
+        ImmutableList.Builder<ReplacingPattern> patternList= ImmutableList.builder();
+        for (String line: IOUtils.readLines(stream, charset)) {
+            line = line.trim();
+            if (!Strings.isNullOrEmpty(line) && !line.startsWith("#")) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: hasContent(line)

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/StripAttachment.java [hunk 22]
```diff
@@ -107,11 +107,7 @@
 
     private boolean decodeFilename = false;
 
-    private Pattern[] replaceFilenamePatterns = null;
-
-    private String[] replaceFilenameSubstitutions = null;
-
-    private Integer[] replaceFilenameFlags = null;
+    private List<ReplacingPattern> filenameReplacingPatterns = null;
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/StripAttachment.java]: = null ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 1 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 2 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 3 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: Already done in next commit.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 4 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 5 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 6 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 7 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: How to get that in java.mail?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 8 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 9 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: That was the previous behaviour, as the charset is part of the configuration; this mailet may be used to set it.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 10 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 11 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 12 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 13 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 14 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: No more

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 15 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 16 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 17 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 18 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: That was previous behaviour, i don't want to break it.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 19 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: Right

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 20 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 21 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/StripAttachment.java (hunk 22 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/StripAttachment.java]: Another mailet marked has not supported.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 12 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: Finally no, as the StringUtils is doing some escaping which is not done by guava:
```java
    /**
     * Splits a string given a pattern (regex), considering escapes.
     * <p> For example considering a pattern "," we have:
     * one,two,three => {one},{two},{three}
     * one\,two\\,three => {one,two\\},{three}
     * <p>
     * NOTE: Untested with pattern regex as pattern and untested for escape chars in text or pattern.
     */
```

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java [hunk 23]
```diff
@@ -88,263 +94,287 @@
...
-        pattern = pattern.trim();
-        if (pattern.length() < 2 && !pattern.startsWith("/") && !pattern.endsWith("/")) throw new MailetException("Invalid parameter value: " + PARAMETER_NAME_SUBJECT_PATTERN);
-        pattern = pattern.substring(1, pattern.length() - 1);
-        String[] patternArray = StringUtils.split(pattern, "/,/");
+
+    private boolean isDebug() {
+        if (Integer.valueOf(getInitParameter("debug", "0")) == 1) {
+            return true;
+        }
+        return false;
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: return Integer.valueOf(getInitParameter("debug", "0")) == 1

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java [hunk 24]
```diff
@@ -88,263 +94,287 @@
...
+            options += Pattern.DOTALL;
+        }
+        return options;
+    }
+
+    private static boolean extractRepeat(String flagsAsString) {
+        if (flagsAsString.indexOf('r') >= 0) {
+            return true;
+        }
+        return false;
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: return flagsAsString.indexOf('r') >= 0;

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java [hunk 25]
```diff
@@ -88,263 +94,287 @@
...
+        if (pieces.length < 3) {
+            throw new MailetException("Invalid expression: " + line);
+        }
+        return new ReplacingPattern(Pattern.compile(pieces[0], extractOptions(pieces[2])), 
+                extractRepeat(pieces[2]), 
+                unescapeSubstitutions(pieces[1]));
+    }
+
+    private static int extractOptions(String optionsAsString) {
+        int options = 0;
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: Could it be something like an EnumSet?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java [hunk 26]
```diff
@@ -19,140 +19,222 @@
...
+        mailet.init(mailetConfig);
 
-        mailet = new ReplaceContent();
-        mci = new FakeMailetConfig("Test", FakeMailContext.defaultContext());
-        mci.setProperty("subjectPatternFile", "#/org/apache/james/mailet/standard/mailets/replaceSubject.patterns");
-        mailet.init(mci);
+        MimeMessage message = new MimeMessage(Session.getDefaultInstance(new Properties()));
+        message.setSubject("una prova");
+        message.setText("Sto facendo una prova di scrittura/ \u00E8 solo una prova.\n"
+                + "Bla bla bla bla.\n");
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: What about using English test sentences?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 23 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 24 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 25 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: Not so simple, as `Pattern.CASE_INSENSITIVE` is an int & the flags parameter in `Pattern.compile` method is the sum of those different ints.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 26 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: text is text whatever the language.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java (hunk 25 above)
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ReplaceContent.java]: ok

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 26 above)
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: true, but in an unknown language is it harder to check substitutions

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 7 above)
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: in getContentType ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 9 above)
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: I think it's very useful but using ReplaceContent for forcing the encoding looks like a hack.
We can change the document (and the behavior) IMO

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 16 above)
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: It's some power of two, it is supposed to be combined with OR operator.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 27]
```diff
@@ -103,35 +109,64 @@ private String unescapeSubstitutions(String substitutions) {
...
+            if (hasContent(line)) {
+                assertPatternSurroundedWithSlashes(line);
                 patternList.add(getPattern(line.substring(1, line.length() - 1)));
             }
         }
         return patternList.build();
     }
 
+    private boolean hasContent(String line) {
+        return !Strings.isNullOrEmpty(line) && !line.startsWith("#");
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: isComment

Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 7 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 9 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 16 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

**Comment by chibenwa** [comment]
:+1:

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+        this.logOwner = logOwner;
+    }
+
+    public String applyPatterns(List<ReplacingPattern> patterns, String text) {
+        for (ReplacingPattern replacingPattern : patterns) {
+            boolean changed;
+            int index = 0;
+            do {
+                changed = false;
+                Matcher matcher = replacingPattern.getMatcher().matcher(text);
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: Could you do some method extraction here ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+                patternList.addAll(getPatternsFromStream(inputStream.get(), fileNameWithCharset.getCharset()));
+            }
+        }
+        return patternList.build();
+    }
+
+    private static class FileNameWithCharset {
+
+        public static FileNameWithCharset from(String fileName) {
+            int charsetOffset = fileName.lastIndexOf('?');
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: extract charset detection

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java [hunk 30]
```diff
@@ -0,0 +1,180 @@
...
+        }
+    }
+
+    private List<ReplacingPattern> getPatternsFromStream(InputStream stream, Charset charset) throws MailetException, IOException {
+        ImmutableList.Builder<ReplacingPattern> patternList= ImmutableList.builder();
+        for (String line: IOUtils.readLines(stream, charset)) {
+            line = line.trim();
+            if (!isComment(line)) {
+                assertPatternSurroundedWithSlashes(line);
+                patternList.add(getPattern(line.substring(1, line.length() - 1)));
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: stripSurroundingSlashes

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java (hunk 28 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/ContentReplacer.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 29 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java (hunk 30 above)
Comment (aduprat) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/PatternExtractor.java]: OK

**Comment by mbaechler** [comment]
👍

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 26 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: Ok, and rebased

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java [hunk 31]
```diff
@@ -129,8 +79,8 @@ public void serviceShouldReplaceSubjectWhenMatching() throws Exception {
     @Test
     public void serviceShouldReplaceBodyWhenMatching() throws Exception {
         mailetConfig.setProperty("subjectPattern", "/prova/PROVA/i/,/a/e//,/o/o/i/");
-        mailetConfig.setProperty("bodyPattern", "/prova/PROVA/i/," + "/a/e//,"
-                + "/o/o/i/,/\\u00E8/e'//," + "/prova([^\\/]*?)ble/X$1Y/im/,"
+        mailetConfig.setProperty("bodyPattern", "/prova/PROVA/i/," + "/a/e/r/,"
+                + "/O/o/r/,/\\u00E8/e'//," + "/prove([^\\/]*?)ble/X$1Y/im/,"
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: Thank you for the translations. You have some remaining prova(ble) in this test.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java [hunk 32]
```diff
@@ -59,41 +59,38 @@ public void getMailetInfoShouldReturnValue() {
...
-        message.setSubject("una prova");
-        message.setText("Sto facendo una prova di scrittura/ \u00E8 solo una prova.\n"
-                + "Bla bla bla bla.\n");
+        message.setSubject("one test");
 
         Mail mail = new FakeMail(message);
         mailet.service(mail);
 
-        assertThat(mail.getMessage().getSubject()).isEqualTo("une PRoVA");
+        assertThat(mail.getMessage().getSubject()).isEqualTo("ane TEsT");
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: s/ane/one/

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java [hunk 33]
```diff
@@ -59,41 +59,38 @@ public void getMailetInfoShouldReturnValue() {
...
-                + "Bla bla bla bla.\n");
+        message.setText("This is one simple test/ \u00E8 one simple test.\n"
+                + "Blo blo blo blo.\n");
 
         Mail mail = new FakeMail(message);
         mailet.service(mail);
 
-        assertThat(mail.getMessage().getContent()).isEqualTo("Sto fecendo une PRoVA di scritture, e' solo une P.\n"
-                + "Q ble ble ble.\n");
+        assertThat(mail.getMessage().getContent()).isEqualTo("This is ane simple TEsT, e' ane simple P.\n"
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: here too

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 32 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: No, that's the goal of the mailet -> changing the text in subject/content.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 33 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: same comment

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 32 above)
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: ok sorry

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 33 above)
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: ok

**Comment by aduprat** [comment]
test this please

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java (hunk 31 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/ReplaceContentTest.java]: done

**Comment by mbaechler** [comment]
:+1:

**Comment by aduprat** [comment]
test this please

**Pull request merged**